cpp/src/streaming/cudf/shuffler.cpp

```diff
@@ -74,29 +176,35 @@ Node shuffler(
         auto partition_map = msg.release<PartitionMapChunk>();
 
         // Make sure that the input chunk's stream is in sync with shuffler's stream.
-        sync_streams(stream, partition_map.stream, event);
-
-        shuffler.insert(std::move(partition_map.data));
+        cuda_stream_join(
+            std::ranges::single_view(stream),
+            std::ranges::single_view(partition_map.stream),
+            &event
+        );
 
-        // Use the highest input sequence number as the output sequence number.
-        sequence_number = std::max(sequence_number, partition_map.sequence_number);
+        shuffler_async.insert(std::move(partition_map.data));
     }
 
     // Tell the shuffler that we have no more input data.
-    std::vector<rapidsmpf::shuffler::PartID> finished(total_num_partitions);
+    std::vector<rapidsmpf::shuffler::PartID> finished(
+        shuffler_async.total_num_partitions()
+    );
     std::iota(finished.begin(), finished.end(), 0);
-    shuffler.insert_finished(std::move(finished));
+    shuffler_async.insert_finished(std::move(finished));
+
+    while (!shuffler_async.finished()) {
+        auto result = co_await shuffler_async.extract_any_async();
+        if (!result.is_valid()) {
+            break;
+        }
 
-    while (!shuffler.finished()) {
-        auto finished_partition = shuffler.wait_any();
-        auto packed_chunks = shuffler.extract(finished_partition);
         co_await ch_out->send(
-            std::make_unique<PartitionVectorChunk>(
-                sequence_number, std::move(packed_chunks), stream
-            )
+            std::make_unique<PartitionVectorChunk>(result.pid, std::move(result.chunks))
         );
```

Comment:
We know how many tasks we must suspend now, I think. So could we change this to:

```
for (int i = 0; i < num_partitions; i++) {
    auto finished = co_await shuffler_async.extract_any_async();
    RAPIDSMPF_EXPECTS(finished.is_valid(), "Something went wrong");
    co_await ch_out->send(std::make_unique<PartitionVectorChunk>(result.pid, std::move(result.chunks)));
}
```